---
title: Errors
slug: en/v2/guides/errors
---

## Introduction

Some error objects may have a `.cause` property. Most of the time, this is of `unknown` type and its value is subject to change without notice in the future.

## General

### SurrealError

`SurrealError` is the base class that most of the error objects thrown explicitly by this SDK inherit from. This class inherits from the JavaScript `Error` class. This error object is rarely thrown alone and is usually thrown by a subclass with a specific meaning. It is recommended to use this error object to easily classify errors caught by `try-catch` syntax as errors explicitly thrown by this SDK.

#### Properties

`.name: "SurrealError"`
:   The error name.

`.message`
:   The error message.

`.stack`
:   Stack trace may be recorded.

`.cause`
:   The cause or context of the error may be set.

### SurrealTypeError

Inherits from: `SurrealError`

This error is thrown when the input value or other validation fails.

#### Properties

`.name: "SurrealTypeError"`
:   The error name.

`.expected`
:   Expected data format.

`.actual`
:   The string representation of the actual value.

#### Resolution

Trace the stack trace to find the cause of the input not being the expected data format.

### SurrealValueError

Inherits from: `SurrealError`

This error is thrown when the input value or other validation fails.

#### Properties

`.name: "SurrealValueError"`
:   The error name.

`.expected`
:   Expected data format.

`.actual`
:   The actual value.

#### Resolution

Trace the stack trace to find the cause of the input not being the expected data format.

### SurrealAggregateError

Inherits from: `SurrealError`

This is an error that aggregates multiple errors or error messages.

#### Properties

`.name: "SurrealAggregateError"`
:   The error name.

`.cause: unknown[]`
:   Each element of the array contains an error or an error message.

### CircularReferenceError

Inherits from: `SurrealError`

This error is thrown when a circular reference is detected. It is mainly thrown when converting JavaScript values to another format when a parent object detects the same object.

#### Properties

`.name: "CircularReferenceError"`
:   The error name.

`.reference: unknown`
:   The value where the circular reference occurred.

For example, this error is thrown in cases like this:

```ts
import { toSurql } from "@tai-kun/surrealdb/utils";

const a = {};
a.a = a;

console.log(a); // <ref *1> { a: [Circular *1] }

toSurql(a); // throws CircularReferenceError
```

#### Resolution

This is a common error that can be thrown even outside the `toSurql` function. You need to carefully debug to see if the object contains itself inside.

### NumberRangeError

Inherits from: `SurrealError`

This error is thrown when a number is out of range.

#### Properties

`.name: "NumberRangeError"`
:   The error name.

`.range`
:   Expected range.

`.actual`
:   The actual value.

`.integer`
:   If `true`, it expects an integer.

#### Resolution

Trace the stack trace to find the cause of the input not being the expected data format.

### UnsupportedRuntimeError

Inherits from: `SurrealError`

This error is thrown when it is determined that an unsupported runtime is being used.

#### Properties

`.name: "UnsupportedRuntimeError"`
:   The error name.

#### Resolution

You must either not use that runtime or use polyfills and thoroughly test.

### UnreachableError

Inherits from: `SurrealError`

This error is thrown when unreachable code is reached. If this error is thrown, you are likely encountering a bug in this SDK.

#### Properties

`.name: "UnreachableError"`
:   The error name.

#### Resolution

You can raise this issue at the following URL:

[https://github.com/tai-kun/surrealdb.js/issues](https://github.com/tai-kun/surrealdb.js/issues)

## Client

### CircularEngineReferenceError

Inherits from: `CircularReferenceError`

This error is thrown when a circular reference occurs between engines when connecting to a database.

#### Properties

`.name: "CircularEngineReferenceError"`
:   The error name.

`.reference: string`
:   The scheme name where the circular reference occurred.

`.seen: string[]`
:   List of scheme names passed through before the circular reference occurred.

For example, this error is thrown in cases like this:

```ts
import { initSurreal } from "@tai-kun/surrealdb";

const { Surreal } = initSurreal({
  engines: {
    http: "https",
    https: "http",
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws CircularEngineReferenceError: Circular engine reference: http,https
```

In the example above, the https protocol engine indicates that it uses the engine set to http, but the http protocol engine tries to use the engine set to https, resulting in a circular reference error.

#### Resolution

Instead of setting the string-specified protocol name to another protocol name's value, replace it with a specific implementation to create an engine:

```ts del={6} ins={7-10}
import { initSurreal } from "@tai-kun/surrealdb";
import HttpEngine from "@tai-kun/surrealdb/http-engine";

const { Surreal } = initSurreal({
  engines: {
    http: "https",
    http: config => new HttpEngine({
      ...config,
      // fetch: <your custom fetch function>
    }),
    https: "http",
  },
  // ...
});
```

### EngineNotFoundError

Inherits from: `SurrealError`

This error is thrown when you try to connect using a protocol that does not have an engine set.

#### Properties

`.name: "EngineNotFoundError"`
:   The error name.

`.scheme: string`
:   The protocol name of the connection destination.

For example, this error is thrown in cases like this:

```ts
import { initSurreal } from "@tai-kun/surrealdb";

const { Surreal } = initSurreal({
  engines: {
    http: config => new HttpEngine(config),
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws EngineNotFoundError: No https scheme engine found.
```

#### Resolution

Check the `engines` property to see if you can create an engine for the protocol of the connection destination endpoint.

```ts ins={6}
import { initSurreal } from "@tai-kun/surrealdb";

const { Surreal } = initSurreal({
  engines: {
    http: config => new HttpEngine(config),
    https: "http",
  },
  // ...
});
```

### ConnectionConflictError

Inherits from: `SurrealError`

This error is thrown when the client attempts to connect to multiple endpoints simultaneously.

#### Properties

`.name: "ConnectionConflictError"`
:   The error name.

`.endpoint1: string`
:   The endpoint that is already connected.

`.endpoint2: string`
:   The other endpoint.

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("ws://localhost:8000");
await db.connect("ws://localhost:1129"); // throws ConnectionConflictError: An attempt was made to connect to ws://localhost:1129/rpc while ws://localhost:8000/rpc was already connected.
```

The `.connect` method appends `/rpc` to the end of the URL if it doesn't end with `/rpc`. Therefore, if the URL path of one endpoint ends in `/rpc`, it may not throw an error, contrary to its appearance:

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.connect("http://localhost:8000/rpc"); // OK!
```

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("http://localhost:8000/rpc");
await db.connect("http://localhost:8000"); // OK!
```

#### Resolution

You can connect to any endpoint by calling the `.close` method before calling the `.connect` method.

```ts ins={6}
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.close();
await db.connect("http://localhost:11298");
```

### MissingNamespaceError

Inherits from: `SurrealError`

This error is thrown when a namespace is not selected before selecting a database. Or, it is also thrown when trying to keep the database selected but leave the namespace unselected.

#### Properties

`.name: "MissingNamespaceError"`
:   The error name.

`.database: string`
:   The database name.

For example, this error is thrown in cases like this:

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.use({
  database: "example",
}); // throws MissingNamespaceError: The namespace must be specified before the database.
```

#### Resolution

Make sure to select the namespace as well when selecting the database.

```ts ins={7}
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.use({
  namespace: "example",
  database: "example",
});
```

### RpcResponseError

Inherits from: `SurrealError`

This error is thrown when the RPC response indicates an error. There is no problem with communication due to the connected protocol or decoding the response body, but it means that SurrealDB cannot handle the RPC request.

#### Properties

`.name: "RpcResponseError"`
:   The error name.

`.id?: string`
:   The ID that identifies the RPC request. The ID always starts with `<method name>_`.

`.code: number`
:   While not explicitly documented in the SurrealDB documentation, it is likely the JSON-RPC error code.

#### Resolution

There are many possible causes, but it is possible that the version of SurrelDB this SDK supports and the version of SurrealDB that processes the RPC request are different.

### QueryFailedError

Inherits from: `SurrealAggregateError`

This error is thrown when the query fails.

#### Properties

`.name: "QueryFailedError"`
:   The error name.

`.cause: string[]`
:   List of errors.

For example, this error is thrown in cases like this:

```ts
await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.query("OUTPUT 'Hello'"); // throws QueryFailedError: Query failed with 1 error(s)
```

#### Resolution

Check if the SurrelQL syntax is correct.

### Closed

Inherits from: `SurrealError`

This error indicates that the connection has been forcibly terminated.

#### Properties

`.name: "Closed"`
:   The error name.

For example, you get this error in cases like this:

```ts
const db = new Surreal();

db.on("<event_name>", ({ signal }) => {
  return new Promise((resolve, reject) => {
    signal.addEventListener("abort", function() {
      console.error(this.reason); // Closed
      reject();
    })

    // ...
  });
});

await db.close({ force: true });
```

## Engine

### EngineError

Inherits from: `SurrealError`

This indicates that the error originated from the engine. It is emitted via the event listener.

#### Properties

`.name: "EngineError"`
:   The error name.

`.fatal: boolean | undefined`
:   Indicates whether this error is fatal.

### HttpEngineError

Inherits from: `EngineError`

This indicates that the error originated from the HTTP engine. Currently, it is only defined and not used.

#### Properties

`.name: "HttpEngineError"`
:   The error name.

### WebSocketEngineError

Inherits from: `EngineError`

This indicates that the error originated from the WebSocket engine. It is used when parsing the response body fails before finding the RPC request ID, when receiving an error event from the WebSocket object, or when the WebSocket connection is disconnected.

This SDK uses `315x` as a custom status code. This range is unassigned in the [IANA](https://www.iana.org/assignments/websocket/websocket.xml#close-code-number).

#### Properties

`.name: "WebSocketEngineError"`
:   The error name.

`.code: number`
:   One of the following status codes:
    - `1000`~`1015`: [MDN Reference](https://developer.mozilla.org/docs/Web/API/CloseEvent/code)
    - `3150`: Indicates that an "error" event was received from the `WebSocket`[^1] instance.
    - `3151`: Indicates that an error occurred in the "open" event handler registered by the WebSocket engine. In most cases, a `StateTransitionError` has occurred during the transition from the connecting state to the connected state.
    - `3152`: Indicates that an error occurred in the "message" event handler registered by the WebSocket engine. There are many possible causes, but it is likely that a `ServerResponseError` or `RpcResponseError` was thrown in the event handler. It is likely caused by a request with invalid RPC parameters. In that case, the method that called the RPC call will timeout (by default) after 5 seconds and fail. In versions prior to SurrealDB v2.0.2, it may timeout for a wider range of causes.
    - `3153`: Indicates that pinging failed. While this may be a temporary error, if you continue to receive it, you may not be able to maintain the connection.
    - `3154`: Indicates that an error occurred in the "close" event handler registered by the WebSocket engine. In most cases, a `StateTransitionError` has occurred during the transition from the disconnecting state to the disconnected state.
:   The following status codes are excluded from the above. This SDK does not treat them as errors.
    - `1000`: Normal disconnect.
    - `1001`: Early disconnection is common.
    - `1004`: Reserved.
    - `1005`: Reserved.
    - `1006`: Reserved.
    - `1015`: Reserved.

^1: `WebSocket` can be a class defined in a global variable by the runtime or a class from `ws`.

#### Resolution

`1002`,`1003`,`1007`...`1011`,`1014`,`3150`,`3151`,`3154`
:   There is probably no automatic way to recover from these status codes. You need to fix the runtime, SurrealDB settings, or hardcoded implementations.

`3152`
:   Check if the arguments passed to the method (that is, the contents of the RPC request) are correct.

`1012`,`1013`,`3153`
:   You may be able to automatically recover using the experimental feature [`autoReconnect()`](/v2/experimental/auto-reconnect).

### StateTransitionError

Inherits from: `SurrealAggregateError`

This error is thrown when the event listener fails to execute during a state transition.

#### Properties

`.name: "StateTransitionError"`
:   The error name.

`.from: string`
:   The state when the state transition started.

`.to: string`
:   The destination state for the transition.

`.fallback: string`
:   The alternative destination for the state transition if it fails. If it is the same value as `.to`, it means that the transition was forced.

For example, this error is thrown in cases like this:

```ts
import { Surreal } from "@tai-kun/surrealdb";

await using db = new Surreal();

db.on("connecting", () => {
  throw new Error("Don't let me connect !!!")
});

await db.connect("http://localhost:8000"); // throws ...
// StateTransitionError: The transition from `3` to `0` failed, falling back to `3`.
//     at <stack trace ...> {
//   cause: [
//     Error: Don't let me connect !!!
//         at <stack trace ...>
//   ],
//   from: 3,
//   to: 0,
//   fallback: 3
// }
```

### ConnectionUnavailableError

Inherits from: `SurrealError`

This error is thrown when trying to send an RPC request in an unconnected state.

#### Properties

`.name: "ConnectionUnavailableError"`
:   The error name.

### ServerResponseError

Inherits from: `SurrealError`

This error is thrown when the response cannot be parsed as a PRC response. This is different from `RpcResponseError`.

#### Properties

`.name: "ServerResponseError"`
:   The error name.

#### Resolution

Normally, this error should not be thrown, but if it is, it is possible that the version of SurrelDB this SDK supports and the version of SurrealDB that processes the RPC request are different.

## CBOR

### CborError

Inherits from: `SurrealError`

This is the class that all CBOR-related errors explicitly thrown by `@tai-kun/surrealdb/cbor` inherit from. It is never thrown directly.

#### Properties

`.name: "CborError"`
:   The error name.

### CborWellFormednessError

Inherits from: `CborError`

#### Properties

This is the class that all CBOR decoding-related errors explicitly thrown by `@tai-kun/surrealdb/cbor` inherit from. It is never thrown directly.

See "Well-Formedness Errors and Examples" in Appendix F of RFC8949 [https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-](https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-).

`.name: "CborWellFormednessError"`
:   The error name.

### CborTooMuchDataError

Inherits from: `CborWellFormednessError`

This indicates that there are unconsumed input bytes remaining.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2)

#### Properties

`.name: "CborTooMuchDataError"`
:   The error name.

#### Resolution

Check if the byte array to be decoded is in the correct CBOR format.

### CborTooLittleDataError

Inherits from: `CborWellFormednessError`

This indicates that the input bytes are incomplete.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4)

#### Properties

`.name: "CborTooLittleDataError"`
:   The error name.

#### Resolution

Check if the byte array to be decoded is in the correct CBOR format.

### CborSyntaxError

Inherits from: `CborWellFormednessError`

Indicates that the input bytes do not match the CBOR encoding format.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6)

#### Properties

`.name: "CborSyntaxError"`
:   The error name.

#### Resolution

Check if the byte array to be decoded is in the correct CBOR format.

### CborMaxDepthReachedError

Inherits from: `CborError`

This error is thrown when the depth of the JavaScript object reaches the maximum value during CBOR encoding or decoding. The depth increases by 1 every time you enter an object or array.

#### Properties

`.name: "CborMaxDepthReachedError"`
:   The error name.

`.maxDepth: number`
:   Maximum depth.

#### Resolution

Relax the upper limit of the optional `.maxDepth` or review the data structure so that the nesting of objects is shallow.

```ts
import CborFormatter from "@tai-kun/surrealdb/cbor-formatter";

const cborFormatter = new CborFormatter({
  encode: {
    maxDepth: 1024,
  },
  decode: {
    maxDepth: 1024,
  },
  // ...
})
```

### CborUnsafeMapKeyError

Inherits from: `CborError`

This error is thrown when an unsafe map key is found during CBOR encoding or decoding.

#### Properties

`.name: "CborUnsafeMapKeyError"`
:   The error name.

`.key: unknown`
:   The map key that was determined to be unsafe.

## JSON

### JsonError

Inherits from: `SurrealError`

This is the error explicitly thrown by `@tai-kun/surrealdb/json`, the JSON formatter.

#### Properties

`.name: "JsonError"`
:   The error name.

### JsonUnsafeMapKeyError

Inherits from: `JsonError`

This error is thrown when an unsafe map key is found during JSON decoding.

#### Properties

`.name: "JsonUnsafeMapKeyError"`
:   The error name.

`.key: unknown`
:   The map key that was determined to be unsafe.
